### PR TITLE
For #12489: Upgrade kotlin, coroutines, compose compiler to latest versions.

### LINKED ISSUE #12489: Upgrade to Kotlin 1.7.10
https://kotlinlang.org/docs/whatsnew17.html

Note: This is currently blocked by Room. The latest version uses Kotlin 1.6.10: https://androidx.tech/artifacts/room/room-ktx/2.4.2. Building with Kotlin 1.7.0 results in errors in the DAO-generated code. 
The latest Room alpha uses 1.6.21.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-20997)
┆Link To Issue: https://mozilla-hub.atlassian.net/browse/FNXV2-20997

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
Fixes #12489

### CODE REVIEW

**Review by jonalmeida — APPROVED** [review]
This looks overall fine to me, with just one comment about a deprecation.

RE landing: let's make sure we do these dependency updates in separate nightly builds. You have another PR that updates separate dependencies. So if there is fallout at runtime, we will be able to see distinguish which group of upgrades may have caused it.

File: components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt [hunk 1]
```diff
@@ -294,20 +294,21 @@ class SystemEngineView @JvmOverloads constructor(
...
                     ErrorType.ERROR_SECURITY_SSL,
-                    error.url
+                    error.url,
                 )?.apply {
                     view.loadUrl(this.uri)
                 }
             }
         }
 
+        @Deprecated("Deprecated in Java")
```
[the remark this answers is not in the document]
Comment (mcarare) [reply components/browser/engine-system/src/main/java/mozilla/components/browser/engine/system/SystemEngineView.kt:304]: Honestly, I was wondering the same. This is the message that is "copied " automatically from the deprecated member. My guess was that if there is a deprecation in Java, this is the standard message that is generated. (The @deprecation annotation in Java does not have a message, in Kotlin it does)

As you can see from your other comment when you automatically copy the deprecation from a Kotlin file it actually copies the message. (It has one to copy from, unlike the Java annotation)

File: components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt [hunk 2]
```diff
@@ -573,7 +573,7 @@ internal class AddonUpdaterWorker(
     private val logger = Logger("AddonUpdaterWorker")
     internal var updateAttemptStorage = DefaultAddonUpdater.UpdateAttemptStorage(applicationContext)
 
-    @Suppress("OverridingDeprecatedMember")
+    @Deprecated("use withContext(...) inside doWork() instead.")
```
[the remark this answers is not in the document]
Comment (mcarare) [reply components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt:576]: I will give this (the replacement) another go. At my first run at this, the resulting behavior was unexpected.

[the remark this answers is not in the document]
Comment (mcarare) [reply components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt]: I added the change, keeping` Dispatchers.Main`.

**Pull request merged**